### DISCUSSION TITLE
REST API PATCH: Why is dict representation allowed for single properties, but not in lists?

### DISCUSSION THREAD

**Discussion by yaiqsa** [netbox-community/netbox#13565] [Help Wanted!]
Hey all,

I was going to create a bugreport, but I suspect that I might just have made a wrong assumption. Let me sketch out my problem:

Netbox version: v3.5.8
Module: Netbox REST API

My goal: Add (or set) tagged vlans on a specific interface.

My API call:
```bash
curl -X PATCH -H "Authorization: Token TOKEN" -H "Content-Type: application/json" -H "Accept: application/json; indent=4" "https://NETBOXINSTANCE/api/dcim/interfaces/5/" --data '{"mode": "tagged", "untagged_vlan":{"vid":1, "site":{"slug":"site-name"}}, "tagged_vlans":[{"vid":"2", "site":{"slug":"site-name"}}]}'
```

Expected result:
Interface with ID 5 now uses tagged mode, with untagged VLAN 1 and tagged VLAN 2

Actual result:
```json
{
    "tagged_vlans": [
        "Incorrect type. Expected pk value, received dict."
    ]
}
```

The interesting thing is though: If I leave out the `tagged_vlans` part, or replace it with `"tagged_vlans":[6]` (in this case 6 being the object ID for the VLAN 2 object), it does work.

So now I'm wondering why I'm allowed to pass a dict representing the VLAN for the `untagged_vlan` option, but not in the list passed to `tagged_vlans`. Or should this actually work, and is this a bug?

Thank you!

**Comment by mtinberg** [comment]
{
    "tagged_vlans": [
        "Incorrect type. Expected pk value, received dict."
    ]
}

That's correct, you just need the id number which is the primary key, not the whole dict with related data, when writing records.
eg

curl -X PATCH [...opts..] --url 'https://netbox/api/dcim/interfaces/1234/ --data '{ "mode": "tagged", "untagged_vlan": 54321, "tagged_vlans": [1234, 4567, 8912] }'

where the number are the ID numbers from ipam.vlans

—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: Yaiqsa ***@***.***>
Sent: Friday, August 25, 2023 7:45 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] REST API PATCH: Why is dict representation allowed for single properties, but not in lists? (Discussion #13565)


Hey all,

I was going to create a bugreport, but I suspect that I might just have made a wrong assumption. Let me sketch out my problem:

Netbox version: v3.5.8
Module: Netbox REST API

My goal: Add (or set) tagged vlans on a specific interface.

My API call:

curl -X PATCH -H "Authorization: Token TOKEN" -H "Content-Type: application/json" -H "Accept: application/json; indent=4" "https://NETBOXINSTANCE/api/dcim/interfaces/5/" --data '{"mode": "tagged", "untagged_vlan":{"vid":1, "site":{"slug":"site-name"}}, "tagged_vlans":[{"vid":"2", "site":{"slug":"site-name"}}]}'

Expected result:
Interface with ID 5 now uses tagged mode, with untagged VLAN 1 and tagged VLAN 2

Actual result:

{
    "tagged_vlans": [
        "Incorrect type. Expected pk value, received dict."
    ]
}

The interesting thing is though: If I leave out the tagged_vlans part, or replace it with "tagged_vlans":[6] (in this case 6 being the object ID for the VLAN 2 object), it does work.

So now I'm wondering why I'm allowed to pass a dict representing the VLAN for the untagged_vlan option, but not in the list passed to tagged_vlans. Or should this actually work, and is this a bug?

Thank you!

**Reply by yaiqsa** [reply]
Hi Mark, thank you for your response!

I understand that I need to pass the vlan IDs for the tagged_vlans list. I just don't understand why it _does_ work to 'describe' the vlan object in the case of the `untagged_vlan` field. Because this does work (in my testing):

curl -X PATCH [...opts..] --url 'https://netbox/api/dcim/interfaces/1234/ --data '{ "mode": "tagged", "untagged_vlan": {"vid":1, "site":{"slug":"site-slug"}}}'

It seems a bit unsystematically that that is allowed, but its not for the untagged_vlan list?

**Comment by mtinberg** [comment]
Maybe you are right and I just haven't retested, looking at your original curl test it looks like in the tagged_vlan list you specified "vid": "2" instead of the number "vid": 2 which can make a difference (number vs string).  Using something other than the primarykey/id may also require that you uniquely and positively identify the resource in a way that the database/orm enforces and guarantees uniqueness, just because you have only one vlan 2 record wouldn't necessarily mean it would work.  I dunno, hth.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by yaiqsa** [reply]
Hm.. For now I can do what I need to do, so I'll just close the discussion. But I or someone else might want to open a ticket in the future to ask if this behaviour is to be expected. (In which case I'm curious for the answer 😉 )
